### PR TITLE
Generate: contrastive search with full optional outputs

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR further massages PT's `contrastive_search` in advance of the conversion to TF. It does the following modifications:
1. Pipes additional outputs that were missing (e.g. when `output_attentions` is `True`)
2. Rewrites part of the input replication to share logic with beam search -- replicating the input for `top_k` candidates is the same as replicating the input for `num_beams`
3. Removes additional redundant/unused operations
4. Because we now have all outputs (see 1), adds the standard suite of tests for a generation method
5. Moves integration tests to the corresponding model folder

All tests passing locally (`RUN_SLOW=1 py.test tests/* -k contrastive -vv`)

After this PR, we can start with the TF conversion.

### CODE REVIEW

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -628,32 +646,29 @@ def _get_decoder_start_token_id(self, decoder_start_token_id: int = None, bos_to
 
     @staticmethod
     def _expand_inputs_for_generation(
-        input_ids: torch.LongTensor,
         expand_size: int = 1,
         is_encoder_decoder: bool = False,
-        attention_mask: Optional[torch.LongTensor] = None,
-        encoder_outputs: Optional[ModelOutput] = None,
```
Comment (gante) [inline src/transformers/generation_utils.py:635]: these input arguments were never used directly, but rather as part of the unpacked `model_kwargs`, so there is no need to have them

File: src/transformers/generation_utils.py [hunk 2]
```diff
@@ -628,32 +646,29 @@ def _get_decoder_start_token_id(self, decoder_start_token_id: int = None, bos_to
 
     @staticmethod
     def _expand_inputs_for_generation(
-        input_ids: torch.LongTensor,
```
Comment (gante) [inline src/transformers/generation_utils.py:631]: This is now optional to share this beam search logic with contrastive search: in contrastive search, we don't need to replicate `input_ids`

File: src/transformers/generation_utils.py [hunk 3]
```diff
@@ -628,32 +646,29 @@ def _get_decoder_start_token_id(self, decoder_start_token_id: int = None, bos_to
...
-        encoder_outputs: Optional[ModelOutput] = None,
+        input_ids: Optional[torch.LongTensor] = None,
         **model_kwargs,
     ) -> Tuple[torch.LongTensor, Dict[str, Any]]:
-        expanded_return_idx = (
-            torch.arange(input_ids.shape[0]).view(-1, 1).repeat(1, expand_size).view(-1).to(input_ids.device)
-        )
-        input_ids = input_ids.index_select(0, expanded_return_idx)
+        if input_ids is not None:
+            input_ids = input_ids.repeat_interleave(expand_size, dim=0)
```
Comment (gante) [inline src/transformers/generation_utils.py:655]: IMO, this way is much more readable than before, hence the change (here and in multiple places in this PR)

It also has a direct TF equivalent (`tf.repeat`), which will make my life easier :)

File: src/transformers/generation_utils.py [hunk 4]
```diff
@@ -1879,13 +1913,14 @@ def contrastive_search(
                 if this_peer_finished_flag.item() == 0.0:
                     break
 
-            # prepare inputs
-            model_kwargs["use_cache"] = True
-            model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
```
Comment (gante) [inline src/transformers/generation_utils.py:1884]: This was outside the `if` below (that only runs in the 1st iteration), and it is only needed in the first iteration -- moved to inside the `if`

File: src/transformers/generation_utils.py [hunk 5]
```diff
@@ -1938,43 +1980,18 @@ def contrastive_search(
...
-                attention_mask = attention_mask.unsqueeze(1).expand(-1, top_k, -1).reshape(-1, attention_mask.size(-1))
-            else:
-                attention_mask = None
+            model_kwargs["past"] = new_key_values
 
-            # encoder-decoder model also contains the `encoder_outputs`
-            if self.config.is_encoder_decoder and "encoder_outputs" in model_inputs:
-                encoder_outputs = model_inputs["encoder_outputs"]
-            else:
-                encoder_outputs = None
```
Comment (gante) [inline src/transformers/generation_utils.py:1969]: This now happens inside `self._expand_inputs_for_generation`, a few lines above

File: src/transformers/generation_utils.py [hunk 6]
```diff
@@ -1938,43 +1980,18 @@ def contrastive_search(
...
-            next_model_inputs = self.prepare_inputs_for_generation(
-                top_k_ids.view(-1, 1),
-                past=past_key_values,
-                attention_mask=attention_mask,
-                use_cache=True,
-                encoder_outputs=encoder_outputs,
-            )
             # compute the candidate tokens by the language model and collects their hidden_states
+            next_model_inputs = self.prepare_inputs_for_generation(top_k_ids.view(-1, 1), **model_kwargs)
```
Comment (gante) [inline src/transformers/generation_utils.py:1994]: With the standardization of operations in the past two PRs, `model_kwargs` holds all the information needed here (just like in the other generate methods)

File: src/transformers/generation_utils.py [hunk 7]
```diff
@@ -628,47 +646,47 @@ def _get_decoder_start_token_id(self, decoder_start_token_id: int = None, bos_to
...
             model_kwargs["encoder_outputs"] = encoder_outputs
+
         return input_ids, model_kwargs
 
     @staticmethod
-    def _update_model_kwargs_for_generation(
-        outputs: ModelOutput, model_kwargs: Dict[str, Any], is_encoder_decoder: bool = False
-    ) -> Dict[str, Any]:
-        # update past
+    def _extract_past_from_model_output(outputs: ModelOutput):
```
Comment (gante) [inline src/transformers/generation_utils.py:675]: This part of the `_update_model_kwargs_for_generation` logic was needed inside `contrastive_search` to make it compatible with all keys for `past`, so I moved it out

File: src/transformers/generation_utils.py [hunk 8]
```diff
@@ -1879,27 +1916,20 @@ def contrastive_search(
...
+                model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
+
                 # encode the given prefix and prepare model inputs; encoder-decoder model process the prefix and save
                 # the `encoder_outputs`
                 outputs = self(
                     **model_inputs, return_dict=True, output_hidden_states=True, output_attentions=output_attentions
                 )
 
-                # past_key_values is required for fast decoding
-                if "past_key_values" not in outputs:
```
Comment (gante) [inline src/transformers/generation_utils.py:1896]: This check was moved below and made more general -- contrastive search is now compatible with other keys for `past`

File: tests/generation/test_generation_utils.py [hunk 9]
```diff
@@ -1336,6 +1405,64 @@ def test_constrained_beam_search_generate_dict_output(self):
...
                 self._check_outputs(output, input_ids, model.config, num_return_sequences=beam_scorer.num_beams)
 
+    def test_contrastive_generate(self):
+        # check `generate()` and `contrastive_search()` are equal
+        for model_class in self.all_generative_model_classes:
+
+            # TODO: fix me. Bloom fails because `past` has a different shape, while FSMT and reformer have a different
+            # cache variable type (and format).
+            if any(model_name in model_class.__name__.lower() for model_name in ["bloom", "fsmt", "reformer"]):
+                return
```
Comment (gante) [inline tests/generation/test_generation_utils.py:1415]: Given the date of addition and usage rate for `fsmt` and `reformer`, I'm assuming we won't fix them -- is this correct?

Comment (sgugger) [reply tests/generation/test_generation_utils.py:1415]: Yes, no need to make an effort for those two.

**Review by sgugger — APPROVED** [review]
Thanks for all the cleanup!

**Pull request merged**